tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/views/source_code/source_code_component.ts

```diff
@@ -0,0 +1,145 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {
+  ChangeDetectionStrategy,
+  Component,
+  ElementRef,
+  Input,
+  SimpleChanges,
+  ViewChild,
+  OnInit,
+} from '@angular/core';
+import {fromEvent, interval} from 'rxjs';
+import {debounce, tap} from 'rxjs/operators';
+
+/** @typehack */ import * as _typeHackRxjs from 'rxjs';
+
+const DEFAULT_CODE_LANGUAGE = 'python';
+const DEFAULT_CODE_FONT_SIZE = 10;
+
+const RESIZE_DEBOUNCE_INTERAVL_MS = 50;
+
+@Component({
+  selector: 'source-code-component',
+  templateUrl: './source_code_component.ng.html',
+  styleUrls: ['./source_code_component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SourceCodeComponent implements OnInit {
+  @Input()
+  lines: string[] | null = null; // TODO(cais): Add spinner for `null`.
+
+  @Input()
+  focusedLineno: number | null = null;
+
+  // TODO(cais): Explore better typing by depending on external libraries.
+  @Input()
+  monaco: any | null = null;
+
+  @ViewChild('codeViewerContainer', {static: true, read: ElementRef})
+  private readonly codeViewerContainer!: ElementRef<HTMLDivElement>;
+
+  private editor: any = null;
+
+  private decorations: string[] = [];
+
+  ngOnInit(): void {
+    // Listen to window resize event. When resize happens, re-layout
+    // monaco editor. Do this with `debounce()` to prevent re-layouting
+    // at too high a rate.
+    fromEvent(window, 'resize')
+      .pipe(
+        debounce(() => interval(RESIZE_DEBOUNCE_INTERAVL_MS)),
+        tap(() => {
+          if (this.editor !== null) {
+            this.editor.layout();
+          }
+        })
+      )
+      .subscribe();
```

Comment:
Since this component can be created and destroyed a lot, please use `takeUntil()` to remove the subscription.

Reply:
Using `takeUntil()` in conjunction with a `Subject` that emits in `ngOnDestroy()` now.